src/ast/nodes/AssignmentExpression.ts

```diff
@@ -26,15 +26,10 @@ export default class AssignmentExpression extends NodeBase {
 		| '**=';
 	right!: ExpressionNode;
 	type!: NodeType.tAssignmentExpression;
-
-	bind() {
-		super.bind();
-		this.left.deoptimizePath(EMPTY_PATH);
-		// We cannot propagate mutations of the new binding to the old binding with certainty
-		this.right.deoptimizePath(UNKNOWN_PATH);
-	}
+	private deoptimized = false;
 
 	hasEffects(context: HasEffectsContext): boolean {
+		if (!this.deoptimized) this.applyDeoptimizations();
```

Comment:
I extracted the check because `hasEffects` and `include` are called quite often and the deoptimizations will only be necessary once. This would save the cost of one function call, but the wins are probably small so we could also move the check back if you prefer.